amazon-connect/callstats-amazon-connect-shim.js

```diff
@@ -1,6 +1,92 @@
 /*! callstats Amazon SHIM version = 1.1.8 */
 
 (function (global) {
+  class VoiceActivityDetection {
+    constructor(stream, callback) {  
+      this.AudioContext = window.AudioContext || window.webkitAudioContext;
+      this.audioContext = new this.AudioContext();
+      this.analyser = this.audioContext.createAnalyser();
+      this.analyser.fftSize = 512;
+      this.analyser.smoothingTimeConstant = 0.1;
+      this.fftBins = new Float32Array(this.analyser.frequencyBinCount);
+      this.mediaStreamSource = this.audioContext.createMediaStreamSource(stream);
+      this.mediaStreamSource.connect(this.analyser);
+      this.isSpeaking = false;
+      this.maxVolumeHistory = [];
+      for (var i=0; i < 10; i++) {
+        this.maxVolumeHistory.push(0);
+      }
+      this.threshold = -50;
+      this.interval = 50;
+      this.callback = callback;
+      this.start();
+    }
+
+    getMaxVolume () {
+      var maxVolume = -Infinity;
+      this.analyser.getFloatFrequencyData(this.fftBins);
+    
+      for(var i=4, ii=this.fftBins.length; i < ii; i++) {
+        if (this.fftBins[i] > maxVolume && this.fftBins[i] < 0) {
+          maxVolume = this.fftBins[i];
+        }
+      };    
+      return maxVolume;
+    }
+
+    start() {
+      var self = this;
+      if (self.timer) {
+        return;
+      }
+      self.timer = setInterval(function() {
+        self.detect();
+      }, self.interval);
+    }
+
+    detect() {
+      var maxVolume = this.getMaxVolume();
+      var totalMaxVolume = 0;
+      if (maxVolume > this.threshold && !this.speaking) {
+        for (var i = this.maxVolumeHistory.length - 3; i < this.maxVolumeHistory.length; i++) {
+          totalMaxVolume += this.maxVolumeHistory[i];
+        }
+        if (totalMaxVolume >= 2) {
```

Comment:
Where does 2 come from?

Reply:
So we're pushing only 1s to the maxVolumeHistory array. Perhaps we rename it to something else like volumeHistory, as there's no max in this data in my opinion.